### PR TITLE
Integrating hvpa-controller into Gardener

### PR DESCRIPTION
**Pull request by ggaurav10**
**What this PR does / why we need it**:
Integrating hvpa-controller into Gardener

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
perf tests are going on, and we will update the results, and then will remove the WIP label.

**Release note**:

```noteworthy operator
The control plane components `kube-apiserver` and `etcd` will be autoscaled horizontally and vertically using [`hvpa-controller`](https://github.com/gardener/hvpa-controller) if the `HVPA` feature gate is enabled. Currently, the update policy for the autoscalers is set to `ScaleUp`, which is intended as a safer first step, however, it might have cost impact. Eventually, the update policy will be changed to `Auto` for `kube-apiserver` first and later for `etcd` in a future release.
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -12,7 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-REGISTRY                           := eu.gcr.io/gardener-project/gardener
+REGISTRY                           := ggaurav10
```
Comment (zanetworker) [inline Makefile]: I guess you forgot to change that right :)?

Comment (ggaurav10) [reply Makefile]: I plan to revert the 2 files - Makefile and VERSION - once all the performance tests are completed. I will also remove the WIP label.
Hope that's ok. :)

Comment (zanetworker) [reply Makefile]: sure, thanks for clarifying :)

**Comment by mvladev** [comment]
I've identified several issues with HVPA which IMHO should be fixed before getting this in:

- [x] https://github.com/gardener/hvpa-controller/issues/31
- [x] https://github.com/gardener/hvpa-controller/issues/32

**Comment by ggaurav10** [comment]
The performance tests results are summarised [here](https://github.com/gardener/hvpa-controller/tree/master/docs/performance-tests/220%20node%20cluster).

@mvladev 
[gardener/hvpa-controller#31](https://github.com/gardener/hvpa-controller/issues/31) is resolved now via [gardener/hvpa-controller#38](https://github.com/gardener/hvpa-controller/pull/38).

However, regarding [gardener/hvpa-controller#32](https://github.com/gardener/hvpa-controller/issues/32), can you please suggest next steps after our last discussion on the github issue page. Thanks.

PS: Removing the WIP label.

File: charts/seed-bootstrap/charts/hvpa/templates/controller-deployment.yaml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    metadata:
+      labels:
+        app: hvpa-controller
+{{ toYaml .Values.labels | indent 8 }}
+    spec:
+      serviceAccountName: hvpa-controller
+      containers:
+      - name: hvpa-controller
+        image: {{ index .Values.global.images "hvpa-controller" }}
+        imagePullPolicy: Always
```
Comment (rfranzke) [inline charts/seed-bootstrap/charts/hvpa/templates/controller-deployment.yaml]:
```suggestion
        imagePullPolicy: IfNotPresent
```

File: charts/seed-bootstrap/charts/hvpa/templates/hvpa-crd.yaml [hunk 3]
```diff
@@ -0,0 +1,972 @@
+apiVersion: apiextensions.k8s.io/v1beta1
+kind: CustomResourceDefinition
+metadata:
+  name: hvpas.autoscaling.k8s.io
```
Comment (rfranzke) [inline charts/seed-bootstrap/charts/hvpa/templates/hvpa-crd.yaml:4]: I think we should rename the API group to not contain `k8s.io` as this is "reserved" by the Kubernetes core/community projects. 

```suggestion
  name: hvpas.autoscaling.gardener.cloud
```

File: charts/seed-bootstrap/charts/hvpa/values.yaml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+global:
+  images:
+    hvpa-controller: image-repository:image-tag
+
+enabled: true
+podAnnotations: {}
+labels:
+  garden.sapcloud.io/role: hvpa
```
Comment (rfranzke) [inline charts/seed-bootstrap/charts/hvpa/values.yaml]:
```suggestion
  gardener.cloud/role: hvpa
```

File: charts/seed-bootstrap/charts/hvpa/values.yaml [hunk 5]
```diff
@@ -0,0 +1,8 @@
+global:
+  images:
+    hvpa-controller: image-repository:image-tag
+
+enabled: true
```
Comment (rfranzke) [inline charts/seed-bootstrap/charts/hvpa/values.yaml]: Make the default `false` here as it's hidden behind a feature gate.

File: charts/seed-bootstrap/values.yaml [hunk 6]
```diff
@@ -115,6 +116,9 @@ alertmanager:
   emailConfigs: []
   storage: 1Gi
 
+hvpa:
+  enabled: true
```
Comment (rfranzke) [inline charts/seed-bootstrap/values.yaml]:
```suggestion
  enabled: false
```

File: charts/seed-controlplane/charts/etcd/templates/etcd-hvpa.yaml [hunk 7]
```diff
@@ -0,0 +1,74 @@
+{{ if .Values.hvpa.enabled }}
+{{- /* .Values.replicas is of type 'float64', so let's cast it to string to have proper types for comparison */}}
+{{- if ne (.Values.replicas | toString) "0" }}
+apiVersion: autoscaling.k8s.io/v1alpha1
+kind: Hvpa
+metadata:
+  #annotations:
```
Comment (rfranzke) [inline charts/seed-controlplane/charts/etcd/templates/etcd-hvpa.yaml]: ?

File: charts/seed-controlplane/charts/etcd/values.yaml [hunk 8]
```diff
@@ -16,5 +16,36 @@ servicePorts:
...
+etcdResources:
+  requests:
+    cpu: 300m
+    memory: 1000M
+  limits:
+    cpu: "4"
+    memory: 30G
+
+hvpa:
+  enabled: true
```
Comment (rfranzke) [inline charts/seed-controlplane/charts/etcd/values.yaml]: `false`

File: charts/seed-controlplane/charts/kube-apiserver/values.yaml [hunk 9]
```diff
@@ -96,3 +112,26 @@ enableBasicAuthentication: true
 #  # The current private key of the service account token issuer. The issuer will sign issued ID
 #  #  tokens with this private key. (Requires the 'TokenRequest' feature gate.)
 #  signingKey: <private key in PEM format>
+
+hvpa:
+  enabled: true
```
Comment (rfranzke) [inline charts/seed-controlplane/charts/kube-apiserver/values.yaml]: `false`

File: pkg/operation/botanist/controlplane.go [hunk 10]
```diff
@@ -34,8 +34,8 @@ import (
 	"github.com/gardener/gardener/pkg/utils/kubernetes/health"
 	"github.com/gardener/gardener/pkg/utils/retry"
 
+	autoscalinghvpa "github.com/gardener/hvpa-controller/api/v1alpha1"
```
Comment (rfranzke) [inline pkg/operation/botanist/controlplane.go]:
```suggestion
	hvpav1alpha1 "github.com/gardener/hvpa-controller/api/v1alpha1"
```

File: pkg/operation/hybridbotanist/controlplane.go [hunk 11]
```diff
@@ -226,6 +239,9 @@ func (b *HybridBotanist) DeployKubeAPIServer() error {
 			apiServer  = b.ShootedSeed.APIServer
 			autoscaler = apiServer.Autoscaler
 		)
+		defaultValues["hvpa"] = map[string]interface{}{
+			"enabled": false,
```
Comment (rfranzke) [inline pkg/operation/hybridbotanist/controlplane.go]: Why do we disable it for shooted seeds?

File: pkg/operation/hybridbotanist/controlplane.go [hunk 12]
```diff
@@ -256,19 +275,52 @@ func (b *HybridBotanist) DeployKubeAPIServer() error {
...
-		defaultValues["apiServerResources"] = map[string]interface{}{
-			"limits": map[string]interface{}{
-				"cpu":    cpuLimit,
-				"memory": memoryLimit,
-			},
-			"requests": map[string]interface{}{
-				"cpu":    cpuRequest,
-				"memory": memoryRequest,
-			},
+		if foundDeployment == false || !hvpaEnabled {
```
Comment (rfranzke) [inline pkg/operation/hybridbotanist/controlplane.go]:
```suggestion
		if !foundDeployment || !hvpaEnabled {
```

File: pkg/operation/hybridbotanist/controlplane.go [hunk 13]
```diff
@@ -256,19 +275,52 @@ func (b *HybridBotanist) DeployKubeAPIServer() error {
...
+					"cpu":    cpuLimit,
+					"memory": memoryLimit,
+				},
+				"requests": map[string]interface{}{
+					"cpu":    cpuRequest,
+					"memory": memoryRequest,
+				},
+			}
+		} else {
+			// Deployment is already created AND is controled by HVPA
```
Comment (rfranzke) [inline pkg/operation/hybridbotanist/controlplane.go]:
```suggestion
			// Deployment is already created AND is controlled by HVPA
```

**Review by rfranzke — CHANGES_REQUESTED** [review]

File: charts/seed-controlplane/charts/etcd/templates/etcd-hvpa.yaml [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+              maxAllowed:
+                memory: 30G
+                cpu: "4"
+              minAllowed:
+                memory: 1000M
+                cpu: 300m
+            - containerName: backup-restore
+              mode: "Off"
+    updatePolicy:
+      updateMode: "ScaleUp"
```
Comment (amshuman-kr) [inline charts/seed-controlplane/charts/etcd/templates/etcd-hvpa.yaml:62]: Can we document for the release note that `ScaleUp` policy is intended as a safer first step but might have cost impact?

File: charts/seed-controlplane/charts/kube-apiserver/templates/kube-apiserver-hvpa.yaml [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+                cpu: "8"
+              minAllowed:
+                memory: 400M
+                cpu: 400m
+            - containerName: vpn-seed
+              mode: "Off"
+            - containerName: blackbox-exporter
+              mode: "Off"
+    updatePolicy:
+      updateMode: "ScaleUp"
```
Comment (amshuman-kr) [inline charts/seed-controlplane/charts/kube-apiserver/templates/kube-apiserver-hvpa.yaml:64]: Can we document for the release note that `ScaleUp` policy is intended as a safer first step but might have cost impact? And that we will work on introducing `Auto` policy for `kube-apiserver`?

File: pkg/operation/hybridbotanist/controlplane.go (hunk 11 above)
Comment (ggaurav10) [reply pkg/operation/hybridbotanist/controlplane.go]: Removed the code

File: charts/seed-bootstrap/charts/hvpa/templates/hvpa-crd.yaml (hunk 3 above)
Comment (ggaurav10) [reply charts/seed-bootstrap/charts/hvpa/templates/hvpa-crd.yaml:4]: Discussed this with @rfranzke offline, and decided to keep it as it is as we plan to move hvpa-controller to autoscaling.k8s.io group
Discussion with the community ongoing.

**Review by amshuman-kr — COMMENTED** [review]
Can we document the possible cost implications of `ScaleUp` update policy for `etcd` and `kube-apiserver`?

**Review by rfranzke — APPROVED** [review]

**Comment by ialidzhikov** [comment]
Ref https://github.com/gardener/gardener/pull/1456#issuecomment-535046429.

**Pull request merged**